x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java

```diff
@@ -185,34 +222,82 @@ public void expandJobs(String expression, boolean allowNoJobs, ActionListener<Qu
+                        jobs.addAll(clusterStateJobs.values());
+                        Collections.sort(jobs, Comparator.comparing(Job::getId));
+                        jobsListener.onResponse(new QueryPage<>(jobs, jobs.size(), Job.RESULTS_FIELD));
+                    }
                 },
                 jobsListener::onFailure
         ));
     }
 
-    private Map<String, Job> expandJobsFromClusterState(String expression, boolean allowNoJobs, ClusterState clusterState) {
+    private Map<String, Job> expandJobsFromClusterState(String expression, ClusterState clusterState) {
         Map<String, Job> jobIdToJob = new HashMap<>();
-        try {
-            Set<String> expandedJobIds = MlMetadata.getMlMetadata(clusterState).expandJobIds(expression, allowNoJobs);
-            MlMetadata mlMetadata = MlMetadata.getMlMetadata(clusterState);
-            for (String expandedJobId : expandedJobIds) {
-                jobIdToJob.put(expandedJobId, mlMetadata.getJobs().get(expandedJobId));
-            }
-        } catch (Exception e) {
-            // ignore
+        Set<String> expandedJobIds = MlMetadata.getMlMetadata(clusterState).expandJobIds(expression);
+        MlMetadata mlMetadata = MlMetadata.getMlMetadata(clusterState);
+        for (String expandedJobId : expandedJobIds) {
+            jobIdToJob.put(expandedJobId, mlMetadata.getJobs().get(expandedJobId));
         }
         return jobIdToJob;
     }
 
+    /**
+     * Get the job Ids that match the given {@code expression}.
+     *
+     * @param expression   the jobId or an expression matching jobIds
+     * @param allowNoJobs  if {@code false}, an error is thrown when no job matches the {@code jobId}
+     * @param jobsListener The jobs listener
+     */
+    public void expandJobIds(String expression, boolean allowNoJobs, ActionListener<SortedSet<String>> jobsListener) {
+        Set<String> clusterStateJobIds = MlMetadata.getMlMetadata(clusterService.state()).expandJobIds(expression);
+        ExpandedIdsMatcher requiredMatches = new ExpandedIdsMatcher(expression, allowNoJobs);
+        requiredMatches.filterMatchedIds(clusterStateJobIds);
+        // If expression contains a group Id it has been expanded to its
+        // constituent job Ids but Ids matcher needs to know the group
+        // has been matched
+        requiredMatches.filterMatchedIds(MlMetadata.getMlMetadata(clusterService.state()).expandGroupIds(expression));
+
+        jobConfigProvider.expandJobsIdsWithoutMissingCheck(expression, false, ActionListener.wrap(
+                jobIdsAndGroups -> {
+                    // Check for duplicate job Ids
+                    jobIdsAndGroups.getJobs().forEach(id -> {
+                        if (clusterStateJobIds.contains(id)) {
+                            jobsListener.onFailure(new IllegalStateException("Job [" + id + "] configuration " +
+                                    "exists in both clusterstate and index"));
+                            return;
```

Comment:
Hmmm....this `return;` does not break out of either lambdas, the rest of the method is executed.
When returning inside a `forEach` lambda like this, it just means "don't execute the rest of this particular iteration of the `forEach` and go straight to the next iteration" Sort of like a `continue;`.
If this is on purpose, please ignore my comment :)

https://repl.it/@ben_w_trent/NestedLambdasOhMY

Reply:
Excellent spot. I added a test and changed to foreach to a plain for loop

Reply:
Sorry about the size of the PR btw, I'm actually resisting the temptation to push more